### PR TITLE
update python version to 3.8

### LINKED ISSUE #4191: Upgrade CI tests to python 3.8
**Description of the issue**

Most of our CI checks (mypy, lint, changed files test, notebook tests, etc.) running on 3.7 - we should upgrade to 3.8, it's faster, newer, more representative of our users' environments.

### PR DESCRIPTION
**Pull request by stubbi**
resolves #4191 

@balopat was this what you had in mind when opening the linked issue?

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -130,8 +130,8 @@ jobs:
     name: Pytest Ubuntu
     strategy:
       matrix:
-        # TODO(#3800): remove 3.6 when Colab switches to 3.7
```
Comment (balopat) [inline .github/workflows/ci.yml:133]: we can remove this TODO for now, we are stuck on 3.6 compatibility until Google upgrades to 3.7.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -130,8 +130,8 @@ jobs:
...
-        python-version: [ '3.6', '3.7', '3.8' ]
+        # TODO(#3800): remove 3.6 when Colab switches to 3.8
+        python-version: [ '3.6', '3.7', '3.8', '3.9' ]
```
Comment (balopat) [inline .github/workflows/ci.yml:134]: This might have to be a separate PR, with potentially a bunch of issues see #3772.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -201,7 +201,7 @@ jobs:
     name: Pytest Windows
     strategy:
       matrix:
-        python-version: [ '3.7', '3.8' ]
+        python-version: [ '3.7', '3.8', '3.9' ]
```
Comment (balopat) [inline .github/workflows/ci.yml:204]: let's introduce 3.9 separately.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -218,7 +218,7 @@ jobs:
     name: Pytest MacOS
     strategy:
       matrix:
-        python-version: [ '3.7', '3.8' ]
+        python-version: [ '3.7', '3.8', '3.9' ]
```
Comment (balopat) [inline .github/workflows/ci.yml:221]: let's introduce 3.9 separately.

**Review by balopat — CHANGES_REQUESTED** [review]
Yes, thank you! Let's remove that TODO. Also if 3.9 passes, we can just close #3772 but if it's a bunch of failures we should do it in a separate PR.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (stubbi) [reply .github/workflows/ci.yml]: Yes, I agree that having it separate would be much better 👍

File: .github/workflows/ci.yml (hunk 1 above)
Comment (stubbi) [reply .github/workflows/ci.yml:133]: removed it

**Review by balopat — APPROVED** [review]
